### PR TITLE
[SofaMiscTopology] Add component TopologyChecker

### LINKED ISSUE #1367: Create TopologyChecker utilities
All tests on the topology should be in a utility out from the container buffer.
This will remove the methods checkTopology() 

See PR #1351 

______________________________________________________

Suggested labels:

### PR DESCRIPTION
**Pull request by epernod**
This PR add a component called TopologyChecker to be added in the scene and link to a MeshTopology / TopologyContainer in order to check the topology either on demand or every end of step.

The methods used to check the topology is similar to the code inside the method  xxxSetTopologyContainer::CheckTopology() 
which will be removed.

**This PR use changes from PR #1593 and won't compile until the previous one is merged**

Fix #1367 

______________________________________________________

By submitting this pull request, I acknowledge that  
**I have read, understand, and agree [SOFA Developer Certificate of Origin (DCO)](https://github.com/sofa-framework/sofa/blob/master/CONTRIBUTING.md#sofa-developer-certificate-of-origin-dco)**.
______________________________________________________

**Reviewers will merge this pull-request only if**  
- it builds with SUCCESS for all platforms on the CI.
- it does not generate new warnings.
- it does not generate new unit test failures.
- it does not generate new scene test failures.
- it does not break API compatibility.
- it is more than 1 week old (or has fast-merge label).

### CODE REVIEW

File: modules/SofaMiscTopology/src/SofaMiscTopology/TopologyChecker.cpp [hunk 1]
```diff
@@ -0,0 +1,787 @@
...
+
+    return result;
+}
+
+
+
+bool TopologyChecker::checkEdgeTopology()
+{
+    bool ret = true;
+    int nbE = m_topology->getNbEdges();
```
Comment (epernod) [inline modules/SofaMiscTopology/src/SofaMiscTopology/TopologyChecker.cpp]: use Size

**Comment by epernod** [comment]
[ci-build][with-all-tests]

**Comment by fredroy** [comment]
Is there any ... test 🤓 ?

**Comment by epernod** [comment]
test the component that will do the tests. Could be possible I think

**Comment by epernod** [comment]
[ci-build][with-all-tests]

**Comment by epernod** [comment]
[ci-build][with-all-tests]

**Comment by epernod** [comment]
[ci-build][with-all-tests]

**Comment by epernod** [comment]
[ci-build][with-all-tests]

**Comment by epernod** [comment]
[ci-build][with-all-tests]

**Comment by epernod** [comment]
@fredroy you wanted some tests.... I have added "some"... I know 2 are failing but I think the mesh is wrong. I will fix that in another PR

File: modules/SofaMiscTopology/SofaMiscTopology_test/TopologyChecker_test.cpp [hunk 2]
```diff
@@ -0,0 +1,1370 @@
...
+        triangles[10] = tmpT;
+        EXPECT_MSG_EMIT(Error);
+        EXPECT_EQ(checker->checkTriangleToEdgeCrossContainer(), false);
+        // restore good container
+        triangles[10] = triangles[0];
+        triangles[0] = tmpT;
+        EXPECT_MSG_NOEMIT(Error);
+        EXPECT_EQ(checker->checkTriangleToEdgeCrossContainer(), true);
+
+        sofa::helper::WriteAccessor< sofa::core::objectmodel::Data<sofa::helper::vector<Topology::Edge> > > edges = topoCon->d_edge;
```
Comment (fredroy) [inline modules/SofaMiscTopology/SofaMiscTopology_test/TopologyChecker_test.cpp]: Advice: for super long line like that, use auto.
And in this case: `auto edges = sofa::helper::getWriteAccessor(topoCon->d_edge);`

File: modules/SofaMiscTopology/src/SofaMiscTopology/TopologyChecker.cpp [hunk 3]
```diff
@@ -0,0 +1,995 @@
...
+    }
+
+    // check edge buffer
+    bool res = true;
+    for (sofa::Index i = 0; i < nbE; ++i)
+    {
+        const auto& edge = my_edges[i];
+        if (edge[0] == edge[1]) {
+            msg_error() << "CheckEdgeTopology failed: edge " << i << " has 2 identical vertices: " << edge;
+            res = false;
```
Comment (fredroy) [inline modules/SofaMiscTopology/src/SofaMiscTopology/TopologyChecker.cpp:132]: Maybe cut the loop once res = false ? (or add the test i < nbE && !res in the for loop)

File: modules/SofaMiscTopology/src/SofaMiscTopology/TopologyChecker.cpp [hunk 4]
```diff
@@ -0,0 +1,995 @@
...
+    {
+        if (ev->getKey() == 'T')
+        {
+            bool res = checkTopology();
+            if (!res)
+                msg_error() << "checkTopology Error!!";
+        }
+    }
+
+    if (simulation::AnimateEndEvent::checkEventType(event) && d_eachStep.getValue())
```
Comment (fredroy) [inline modules/SofaMiscTopology/src/SofaMiscTopology/TopologyChecker.cpp:975]: empty statement if animateEnd, so I guess you can remove this

File: modules/SofaMiscTopology/src/SofaMiscTopology/TopologyChecker.h [hunk 5]
```diff
@@ -0,0 +1,159 @@
...
+public:
+    /// bool to check topology at each step.
+    Data<bool> d_eachStep;
+
+    /// Link to be set to the topology container in the component graph.
+    SingleLink<TopologyChecker, sofa::core::topology::BaseMeshTopology, BaseLink::FLAG_STOREPATH | BaseLink::FLAG_STRONGLINK> l_topology;
+
+
+protected:
+    core::topology::BaseMeshTopology* m_topology;
```
Comment (fredroy) [inline modules/SofaMiscTopology/src/SofaMiscTopology/TopologyChecker.h]: a SPtr instead of a raw pointer ?

**Comment by fredroy** [comment]
> 
> 
> @fredroy you wanted some tests.... I have added "some"... I know 2 are failing but I think the mesh is wrong. I will fix that in another PR

Could you identify them in the code (in a comment, maybe add the url or the number of this PR) 
In case you dont have time to solve them or something like that; otherwise we wont know when these tests had failed 😏

File: modules/SofaMiscTopology/src/SofaMiscTopology/TopologyChecker.cpp (hunk 3 above)
Comment (epernod) [reply modules/SofaMiscTopology/src/SofaMiscTopology/TopologyChecker.cpp:132]: yes, I hesitated, the idea was to catch all the errors... but it quickly become a mess..

File: modules/SofaMiscTopology/SofaMiscTopology_test/TopologyChecker_test.cpp (hunk 2 above)
Comment (epernod) [reply modules/SofaMiscTopology/SofaMiscTopology_test/TopologyChecker_test.cpp]: thanks, done!

File: modules/SofaMiscTopology/src/SofaMiscTopology/TopologyChecker.cpp (hunk 4 above)
Comment (epernod) [reply modules/SofaMiscTopology/src/SofaMiscTopology/TopologyChecker.cpp:975]: in fact it was not supposed to be empty

**Comment by epernod** [comment]
> 
> 
> > @fredroy you wanted some tests.... I have added "some"... I know 2 are failing but I think the mesh is wrong. I will fix that in another PR
> 
> Could you identify them in the code (in a comment, maybe add the url or the number of this PR)
> In case you dont have time to solve them or something like that; otherwise we wont know when these tests had failed 😏

In fact the checker had a bug. All the tests are now working. Thanks for the review, the tests were worth it!

**Pull request merged**